localstack/services/dynamodb/dynamodb_listener.py

```diff
@@ -133,23 +150,84 @@ def return_response(self, method, path, data, headers, response):
                 record['dynamodb']['SizeBytes'] = len(json.dumps(updated_item))
         elif action == '%s.BatchWriteItem' % ACTION_PREFIX:
             records = []
-            for table_name, requests in data['RequestItems'].items():
-                for request in requests:
+            i = 0
+            for table_name in sorted(data['RequestItems'].keys()):
+                for request in data['RequestItems'][table_name]:
                     put_request = request.get('PutRequest')
                     if put_request:
+                        existing_item = self._thread_local('existing_items')[i]
                         keys = dynamodb_extract_keys(item=put_request['Item'], table_name=table_name)
                         if isinstance(keys, Response):
                             return keys
                         new_record = clone(record)
-                        new_record['eventName'] = 'INSERT'
+                        new_record['eventName'] = 'INSERT' if not existing_item else 'MODIFY'
                         new_record['dynamodb']['Keys'] = keys
                         new_record['dynamodb']['NewImage'] = put_request['Item']
+                        if existing_item:
+                            new_record['dynamodb']['OldImage'] = existing_item
+                        new_record['eventSourceARN'] = aws_stack.dynamodb_table_arn(table_name)
+                        records.append(new_record)
+                    delete_request = request.get('DeleteRequest')
+                    if delete_request:
+                        keys = delete_request['Key']
+                        if isinstance(keys, Response):
+                            return keys
+                        new_record = clone(record)
+                        new_record['eventName'] = 'REMOVE'
+                        new_record['dynamodb']['Keys'] = keys
+                        new_record['dynamodb']['OldImage'] = self._thread_local('existing_items')[i]
                         new_record['eventSourceARN'] = aws_stack.dynamodb_table_arn(table_name)
                         records.append(new_record)
+                    i += 1
+        elif action == '%s.TransactWriteItems' % ACTION_PREFIX:
```

Comment:
nit: Looks like this method is getting pretty long - would that be a good opportunity to break out the new code into small util functions?

Reply:
Refactoring the additions.